### PR TITLE
Allow using a folder path in WebCmdlet's OutFile parameter

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary
In Invoke-WebRequest and Invoke-RestMethod `-Outfile` only works if you specify a file path
After this PR:

- If you specify an existing directory, the file will be placed by its server-designated name there.
- If you specify a file path, that path is saved to as-is; a file path is either:
      - the path to an existing file (which is then quietly replaced, as before)
      - the path to an existing directory followed by the name of a new file to be created

@mklement0
The filename is taken from the last segment of the Uri

Problems: 

- it currently doesn't work with `-Resume` and I'm not sure it can
- I don't know how to make this an experimental feature

## PR Context
#11671

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/9788
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs [hunk 1]
```diff
@@ -56,6 +57,25 @@ internal override void ProcessResponse(HttpResponseMessage response)
...
+                        if (enumerator.MoveNext())
+                        {
+                            // Get file name from Content-Disposition header if present
+                            OutFile = Path.Combine(OutFile, Path.GetFileName((string)enumerator.Current));
+                        }
+                    }
+                    else
+                    {
+                        // Get file name from last segment of Uri
+                        OutFile = Path.Combine(OutFile, System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:75]: I don't value. It is better to generate an error.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:75]: I think you're wrong, I have tested some websites and they rarely (never) use the `Content-Disposition` header. So it would always generate an error.
Tested downloads: Firefox, Chrome, 7zip, Adobe Reader

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:75]: In this case there is no point in analyzing Content-Disposition at all :-)

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: Removed

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: LGTM. Please update the PR description.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: Done. What should we do about the problem with `-Resume` ?

**Comment by CarloToso** [comment]
The test error looks like a random failure unrelated to this PR

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs [hunk 2]
```diff
@@ -141,6 +141,13 @@ internal override void ProcessResponse(HttpResponseMessage response)
             }
             else if (ShouldSaveToOutFile)
             {
+                // Check if OutFile is a folder
+                if (Directory.Exists(OutFile))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:145]: `OutFile` can be provider-qualified path or relative path, ex. `FileSystem::~\out.zip`.
I think we should use `QualifiedOutFile`.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs]: I changed it, but it was working correctly with `FileSystem::~\out.zip` and `FileSystem::~\out`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs [hunk 3]
```diff
@@ -141,6 +141,13 @@ internal override void ProcessResponse(HttpResponseMessage response)
             }
             else if (ShouldSaveToOutFile)
             {
+                // Check if OutFile is a folder
+                if (Directory.Exists(QualifiedOutFile))
+                {
+                    // Get file name from last segment of Uri
+                    OutFile = Path.Combine(OutFile, System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:148]: Now we do path resolving twice since QualifiedOutFile is not cache result.
I'd expect we cache QualifiedOutFile in local variable, modify it here and use below.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:148]: And we need tests for the new scenarios.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs [hunk 4]
```diff
@@ -141,7 +141,16 @@ internal override void ProcessResponse(HttpResponseMessage response)
             }
             else if (ShouldSaveToOutFile)
             {
-                StreamHelper.SaveStreamToFile(baseResponseStream, QualifiedOutFile, this, response.Content.Headers.ContentLength.GetValueOrDefault(), _cancelToken.Token);
+                string qualifiedOutFile = QualifiedOutFile;
+
+                // Check if OutFile is a folder
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:146]: Please remove obvious comment.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 5]
```diff
@@ -544,6 +544,13 @@ internal virtual void ValidateParameters()
                 ErrorRecord error = GetValidationError(WebCmdletStrings.OutFileMissing, "WebCmdletOutFileMissingException", nameof(Resume));
                 ThrowTerminatingError(error);
             }
+
+            // OutFile must not be a directory to use Resume.
+            if (Resume.IsPresent && Directory.Exists(OutFile))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:549]: I'd expect we use resolved path here too.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs [hunk 6]
```diff
@@ -56,7 +56,16 @@ internal override void ProcessResponse(HttpResponseMessage response)
 
             if (ShouldSaveToOutFile)
             {
-                StreamHelper.SaveStreamToFile(responseStream, QualifiedOutFile, this, response.Content.Headers.ContentLength.GetValueOrDefault(), _cancelToken.Token);
+                string qualifiedOutFile = QualifiedOutFile;
+
+                // Check if OutFile is a folder
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:61]: Please remove obvious comment.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 5 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Fixed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 7]
```diff
@@ -544,6 +544,13 @@ internal virtual void ValidateParameters()
                 ErrorRecord error = GetValidationError(WebCmdletStrings.OutFileMissing, "WebCmdletOutFileMissingException", nameof(Resume));
                 ThrowTerminatingError(error);
             }
+
+            // OutFile must not be a directory to use Resume.
+            if (Resume.IsPresent && Directory.Exists(QualifiedOutFile))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:549]: 1. If we resolve the path we should cache it here and use later.
2. It seems it is not correct check since we could get the file name from URL.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:549]: 1. I agree with you, I'll do it later
2. I think it's correct, this is the logic:
- if -Outfile points to a file, old logic you can use resume
- if -Outfile points to a folder (Directory.Exists(QualifiedOutFile)) you can't use the resume switch and the filename is taken later from the url

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: @iSazonov I think I need a hand with caching the resolved path

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 8]
```diff
@@ -397,6 +397,10 @@ public virtual string CustomMethod
 
         #endregion Virtual Properties
 
+        internal string _qualifiedOutFile;
+
+        //internal bool _shouldSaveToOutFile;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:402]: Please remove and look CodeFactor issues.

**Comment by iSazonov** [comment]
@CarloToso We need new tests for null, empty, folder.

**Comment by CarloToso** [comment]
@iSazonov I added the tests for null and empty -OutFile to #19044

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 7 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: > * if -Outfile points to a folder (Directory.Exists(QualifiedOutFile)) you can't use the resume switch (throws an error) and the filename is taken later from the url

If we take a file name from URL we can use Resume, can't we?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 9]
```diff
@@ -570,6 +570,15 @@ internal virtual void ValidateParameters()
...
             }
+
+            _qualifiedOutFile = ShouldSaveToOutFile ? QualifiedOutFile : null;
+
+            // OutFile must not be a directory to use Resume.
+            if (Resume.IsPresent && Directory.Exists(_qualifiedOutFile))
+            {
+                ErrorRecord error = GetValidationError(WebCmdletStrings.ResumeNotFilePath, "WebCmdletResumeNotFilePathException", _qualifiedOutFile);
+                ThrowTerminatingError(error);
+            }
```
[the remark this answers is not in the document]
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:581]: It doesn't work because, we get the last segment of the url after the redirections.
Example:
`$x = Invoke-WebRequest "https://download.mozilla.org/?product=firefox-stub&os=win"`
`$x.BaseResponse.RequestMessage.RequestUri.Segments[-1] --> Firefox%20Installer.exe`
`?product=firefox-stub&os=win  !=  Firefox%20Installer.exe [DESIRED RESULT]`

We can't obtain the correct name until after the request

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx [hunk 10]
```diff
@@ -200,6 +200,9 @@
   </data>
   <data name="ReadResponseProgressStatus" xml:space="preserve">
     <value>Downloaded: {0} of {1}</value>
+  </data>
+   <data name="ResumeNotFilePath" xml:space="preserve">
+    <value>Path '{0}' resolves to a directory. The Resume switch can only be used on paths including a file name.</value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx:205]: Please reword. 
Something like:
```
    <value>The Resume switch can only be used if OutFile targets to a file but it is resolved to a directory: {0}.</value>
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs [hunk 11]
```diff
@@ -140,8 +140,14 @@ internal override void ProcessResponse(HttpResponseMessage response)
                 }
             }
             else if (ShouldSaveToOutFile)
-            {
-                StreamHelper.SaveStreamToFile(baseResponseStream, QualifiedOutFile, this, response.Content.Headers.ContentLength.GetValueOrDefault(), _cancelToken.Token);
+            {   
+                if (Directory.Exists(_qualifiedOutFile))
+                {
+                    // Get file name from last segment of Uri
+                    _qualifiedOutFile = Path.Combine(_qualifiedOutFile, System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]));
```
[the remark this answers is not in the document]
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs]: Done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs [hunk 12]
```diff
@@ -56,7 +56,13 @@ internal override void ProcessResponse(HttpResponseMessage response)
 
             if (ShouldSaveToOutFile)
             {
-                StreamHelper.SaveStreamToFile(responseStream, QualifiedOutFile, this, response.Content.Headers.ContentLength.GetValueOrDefault(), _cancelToken.Token);
+                if (Directory.Exists(_qualifiedOutFile))
+                {
+                    // Get file name from last segment of Uri
+                    _qualifiedOutFile = Path.Combine(_qualifiedOutFile, System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]));
```
[the remark this answers is not in the document]
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: Done

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx (hunk 10 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx]: Done

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@CarloToso Please open new issue in Docs repo and add the reference in the PR description.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs (hunk 1 above)
Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) may not be used _often_ (I can't speak to that), but it is definitely used, including on GitHub:

E.g., in an archive download for a given project, such as `https://github.com/mklement0/fileicon/archive/stable.zip`, where using the last URL segment wouldn't even tell you the name of the product, the `Content-Disposition` field provides a helpful file name:

```powershell
# -> 'attachment; filename=fileicon-stable.zip'
(Invoke-WebRequest -Method Head  https://github.com/mklement0/fileicon/archive/stable.zip).Headers['Content-Disposition']
```

Both `curl` and `wget` support `Content-Disposition`, albeit on an _opt-in_ basis (`-J` and `--content-disposition`)

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: @mklement0 Please describe the desired behaviour and I will implement it, I removed the check for `Content-Disposition` only because @iSazonov requested it

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: Thanks, @CarloToso. 

My suggestion is as follows - though I wonder if it requires a wider discussion:

I suggest also making the feature _opt-in_, given the following warning from `man curl` with respect to its opt-in (`-J` / `--remote-header-name`):

> WARNING: Exercise judicious use of this option, especially on
          Windows. A rogue server could send you the name of a DLL or
          other file that could be loaded automatically by Windows or some
          third party software.

That requires a _new switch_, say `-UseSuggestedFileName`, which would have to report an error if the `-OutFile` argument targets a _file_.

There's also a _discoverability_ problem - also with the use-last-URI-segment approach, given that _redirections_ may be involved, and that the new default behavior uses the _redirected-to_ URI's last segment:

* How does the user learn what file name was ultimately used (short of examining the target directory for the most recently added file)?
  * At the very least, `-Verbose` should include this information.
  * `-PassThru` should use an ETS instance member to decorate the passed-through `Microsoft.PowerShell.Commands.BasicHtmlWebResponseObject` instance with a `.FullName` property that reflects the output file's full path.

As for how to parse `Content-Disposition` and fallback behavior:

* Based on the [previously linked description](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition), the C# equivalent of the following is probably good enough in practice, though the recommendation is to give precedence to `filename*` (sic) entries:

   * `<value-of-Content-Disposition-field(s)> -replace '^.*\bfilename\*?="?([^;"]+)?.*', '$1'`
   *  If there's no `Content-Disposition` field or if the file-name extraction fais, fall back to the last-URI-segment approach

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs]: Maybe you could move this discussion to issue https://github.com/PowerShell/PowerShell/issues/11671

**Comment by CarloToso** [comment]
```ps1
#firefox download
curl "https://download.mozilla.org/?product=firefox-stub&os=win&lang=it" -L -O
#--> _product=firefox-stub&os=win&lang=it
Invoke-WebRequest "https://download.mozilla.org/?product=firefox-stub&os=win&lang=it" -OutFile .\
#--> Firefox Installer.exe (BETTER)

#github download
curl "https://github.com/mklement0/ttab/archive/stable.zip" -L -O
#--> stable.zip (BETTER)
Invoke-WebRequest "https://github.com/mklement0/ttab/archive/stable.zip" -OutFile .\
#--> stable
```
@mklement0 It seems to me that both approaches have some issues... What should we do?

**Comment by mklement0** [comment]
@CarloToso, yes, both approaches are defensible.

* The `curl` approach has one advantage: it is obvious to the user what the output file name will be - though it may not be a meaningful one.

* Your current approach may provide a more meaningful filename, but the user will have to investigate the output directory to find out what filename was used.

---

The `curl` approach is the simplest one (but note that `curl` also offers `-J` to honor a `Content-Disposition` filename).

If we stick with what you have now, we should offer a way to discover the filename via (a) `-Verbose` output and (b) via the object returned when you use `-PassThru` (in the simplest form with an ETS instance property containing a `[System.IO.FileInfo]` instance describing the output file) - this is why I asked for this aspect to be decided on first - see https://github.com/PowerShell/PowerShell/issues/11671#issuecomment-1446282164 

The latter will also become important if and when we implement `Content-Disposition` support later.

(Curiously `curl -JO` also doesn't report the filename ultimately used, and while you can glean it from the `-v` (verbose) output, it's buried in a lot of other info; I think we can do better here).

**Comment by CarloToso** [comment]
@mklement0 I'm slowly working on addressing your suggestions about `-PassThru`, I created `WebResponseHelper.GetOutFilePath` as a stepping stone. I became aware of the problem you just described some hours ago and I'm working on a solution for a follow up PR. My current Roadmap is:
- This PR --> first implementation, it only works for files using the last segment of the uri (fails for `http://example.org`), filename shown if `-Verbose`

One or Two other PR:
- Add Content-Disposition support, choose each time between curl method and my method for filenames
- Handle `http://example.org` case
- Add warnings (like curl) each time we choose a folder as download target, discovery of the output filename via -PassThru, add more tests

I don't think I should add more commits to this PR

**Comment by mklement0** [comment]
Thank you, @CarloToso, that makes sense, except I'm unclear on what you mean by:

> choose each time between curl method and my method for filenames

> Add warnings (like curl) each time we choose a folder as download target

**Comment by CarloToso** [comment]
@mklement0 
1) Each time choose what filename works better between:

- Uri.AbsoluteUri.Split("/")[^1] (sanitized) [curl]
- response.RequestMessage.RequestUri.Segments[^1] [my method]

2) Add a warning message when the user chooses a folder path with `-OutFile`, you said it could be unsafe for the user 

> WARNING: Exercise judicious use of this option, especially on
Windows. A rogue server could send you the name of a DLL or
other file that could be loaded automatically by Windows or some
third party software.

**Comment by mklement0** [comment]
Thanks for clarifying , @CarloToso.

Just food for thought, we can defer further discussion to your future PRs:

Re 1: How would you decide in cases where both methods yield a viable filename?

Re 2: Personally, I think warning _in the documentation_ is sufficient (which is what `curl` does).

**Comment by CarloToso** [comment]
@iSazonov @SteveL-MSFT Can we merge this? So I con continue working on the name from Content-Disposition feature

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Review by SteveL-MSFT — APPROVED** [review]
